docs/exceptions.md

```diff
@@ -61,9 +69,18 @@ scenarios.
 Another example is parsing user input. We might expect the user to enter a
 number, but we have to consider the case that the user enters something else.
 
+*Comment:* a user giving input which does not parse isn't an error or an
+exception, it's a completely normal thing, well-described by the typical
+`Either ParseError` type. What we're dealing with here is a partial function
+from the input type to the type of the parsed thing, which we make total via the
+sum type. This has nothing at all to do with exception handling.
```

Comment:
> Call it what you want — exception handling, error handling, "completely normal thing" handling, but it's still a secondary codepath that is normally implemented with ExceptT, CPS-d ExceptT, IO exceptions, or similar. This is the sort of thing that this document is concerned with.

What we call it is important, apparently. I object to how liberal the characterisation of "exception" or "regular error" is within this document. It makes truly normal things seem convoluted when they're not. And what, exactly, is a "secondary codepath"?

Reply:
> What we call it is important, apparently

Well, yes, given that most of your criticism is based on a misinterpretation of what's being said.

> And what, exactly, is a "secondary codepath"?

Do you expect me to come up with a formal definition? I think you have an intuitive understanding of what this means, it's something that is secondary to the logic of the computation and interrupts its normal execution. We usually represent these with `ExceptT` or `IO` because that's the easiest way to interrupt the computation and reach the handler for it.

Reply:
> Do you expect me to come up with a formal definition?

No, I'm challenging the document. "Secondary codepath" is vague and dubious, and I really don't expect a formal definition. Maybe I have an intuitive understanding of what it means, but does it match up with yours and everybody else's? Probably not.

Dealing with a `Nothing` from a map lookup is not a "secondary" thing, it's exactly as important as the  case branch for `Just`. Neither branch should be considered "secondary". Another hole to poke: what is "normal execution"? Surely, in a pure function, *everything* in its definition is "normal", no?

Reply:
We already covered this: https://github.com/input-output-hk/cardano-sl/pull/2492#issuecomment-364615107